### PR TITLE
jasmine: Make Matchers generic 

### PR DESCRIPTION
**Pull request by lukas-zech-software**
This PR introduces a generic parameter on the `Matchers` object and all `expect` functions.

It asserts that all `MatchFunctions` can only be called with a `expected`parameter that has a compatible type to the `actual` parameter provided in `expect`.

Additionally `jasmine.objectContaining` now returns a `Partial` of the provided type to assert that the provided properties match the ones from `expected`

I also fixed a bunch of lint errors, therefore this PR seems a bit cluttered. Have a look at the first and second commit of this PR to see the [functional changes only with whitespaces hidden](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/14875/commits/e45de33a3f658465e6c96a6b686eea9afc49757e?w=1)

- [x] Make your PR against the `master` branch.
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.) 
*(Tested on our expansive code base and it broke nothing but revealed some hard to find errors thanks to new type asserts)*
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `tsc` without errors.
- [x] Run `npm run lint package-name` if a `tslint.json` is present. 
*(I fixed most, but some where I am unsure remain)*

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://jasmine.github.io/edge/introduction#section-Matchers
 *(Matching against an incompatible type doesn't make sense)*
- [x] Increase the version number in the header if appropriate. 
*(Only improved current typings)*

### CODE REVIEW

**Comment by lukas-zech-software** [comment]
@cnishina Could you please have a look at the changes as well

Due to the augmentation of the `Matcher` interface in the package `jasminewd2` i needed to add my changes to this packages as well to make the CI pass.

As I have no real idea what the packages does, I tried to mimic the already present code.

 please review the changes I made if they make sense

**Comment by cnishina** [comment]
@sjelin should do this review.

**Comment by lukas-zech-software** [comment]
@borisyankov @theodorejb @davidparsson @sjelin 

anyone up for a review? :octocat:

**Comment by minestarks** [comment]
This looks good to me but I know nothing about jasmine. I'll merge in a couple of days if no response from @sjelin @davidparsson @theodorejb @borisyankov

**Comment by lukas-zech-software** [comment]
@noomorph 
@bcaudan 
@cjbarth 
@LOZORD

any contributor to have a look at my changes?

I'd really like to have those changes reviewed by someone else before they are merged

File: jasmine/jasmine-tests.ts [hunk 1]
```diff
@@ -662,28 +662,34 @@ describe("jasmine.any", function () {
...
             bar: "baz"
         };
     });
 
-    it("matches objects with the expect key/value pairs", function () {
-        expect(foo).toEqual(jasmine.objectContaining({
-            bar: "baz"
+    it("matches objects with the expect key/value pairs", () => {
+        expect(foo).toEqual(jasmine.objectContaining<fooType>({
+            bar: ''
```
Comment (LOZORD) [inline jasmine/jasmine-tests.ts]: It might be good to leave a comment here explaining why
```ts
{bar: 'baz'} =~ {bar: ''}
```
as implied by the test.

File: jasminewd2/index.d.ts [hunk 2]
```diff
@@ -44,6 +44,13 @@ declare namespace jasmine {
     toThrowError(expected?: new (...args: any[]) => Error | Promise<new (...args: any[]) => Error>, message?: string | RegExp | Promise<string | RegExp>): Promise<void>;
   }
 
+  interface ArrayLikeMatchers<T extends ArrayLike<T>> extends Matchers<T> {
```
Comment (LOZORD) [inline jasminewd2/index.d.ts]: The `T extends ArrayLike<T>` syntax is a little confusing for me since it appears to be self-referential. Could you please explain why it can't just be `ArrayLike<T>`?

**Review by LOZORD — COMMENTED** [review]
Thanks for your work on this! Your changes will be a great improvement for Jasmine type users. I've left a few minor comments. Everything looks good for the most part.

**Comment by noomorph** [comment]
@lukas-zech-software , wow! :+1:  I'll take a look how it works on a few big Jasmine-dependent projects and will let you know if there any overlooked points today or probably tomorrow. Does it work for you?

**Comment by lukas-zech-software** [comment]
@noomorph 
I have it running it 4 projects and so far it revealed a few test where we actually used wrong types in the `expects`.

Didn't have any other problems so far, but I think our tests (around 1500) are just to narrow minded to  make sure it works for everyone, so every extra tests are heavily welcome

File: jasmine/index.d.ts [hunk 3]
```diff
@@ -24,8 +24,9 @@ declare function afterEach(action: (done: DoneFn) => void, timeout?: number): vo
 declare function beforeAll(action: (done: DoneFn) => void, timeout?: number): void;
 declare function afterAll(action: (done: DoneFn) => void, timeout?: number): void;
 
-declare function expect(spy: Function): jasmine.Matchers;
-declare function expect(actual: any): jasmine.Matchers;
+declare function expect<T extends ArrayLike<T>>(actual: ArrayLike<T>): jasmine.ArrayLikeMatchers<T>;
+declare function expect<T>(actual: T): jasmine.Matchers<T>;
+declare function expect(spy: Function): jasmine.Matchers<any>;
```
Comment (lukas-zech-software) [inline jasmine/index.d.ts]: i moved the function up to be the first

File: jasmine/index.d.ts [hunk 4]
```diff
@@ -113,18 +124,14 @@ declare namespace jasmine {
         withMock(func: () => void): void;
     }
 
-    interface CustomEqualityTester {
-        (first: any, second: any): boolean;
-    }
+    type CustomEqualityTester = (first: any, second: any) => boolean;
```
Comment (lukas-zech-software) [inline jasmine/index.d.ts]: I have not used `CustomEqualityTester` myself so far, so I'm unsure about its usage.

In the example it doesn't make much sense as the function tests for the type of the parameter during runtime to do different  equality tests, so `any` would be the best type here *any*way (no pun intended 😄 )

If both parameters to compare are of a different type, we would need two generic parameters which would complicated matter quickly

If this makes sense in real world usage, I would  add it in a different PullRequest

File: jasmine/jasmine-tests.ts (hunk 1 above)
Comment (lukas-zech-software) [inline jasmine/jasmine-tests.ts]: I add a few comments to explain the usage

File: jasminewd2/index.d.ts (hunk 2 above)
Comment (lukas-zech-software) [inline jasminewd2/index.d.ts]: The generic notation expects you to start with a name for the generic parameter
so you have to start with `<T ...`

The next part tells the compiles that the parameter `T` must be compatible with `ArrayLike<T>`

So basically it reads like:
> This interface is generic for a parameter named `T` which is of type `ArrayLike<T>`


Using just `ArrayLike<T>` would not give  the parameter any name and the compiler complains that he does not know `T`

**Comment by lukas-zech-software** [comment]
btw, am I supposed to rebase the changes so that the PR creates only a single commit on the master?

**Comment by noomorph** [comment]
@lukas-zech-software , I can't tell for sure, but personally, I would prefer to see the commits squashed into a single one since they are about the same thing. Just my opinion.

File: jasminewd2/index.d.ts (hunk 2 above)
Comment (minestarks) [inline jasminewd2/index.d.ts]: +1 to what @LOZORD said. This looks a bit suspicious. When you have ```T extends ArrayLike<T>```, then your constraint is essentially saying the type must be an ArrayLike<T> _in addition to_ still being the element type, ```T```. A regular array e.g. ```[1,2,3]``` will not meet this constraint. The only type that will is one that's an array of itself, like:

```ts
class SelfArray {
    length: number;
    [index: number]: SelfArray;
    someElementProperty: string
}
```

Is this what you want? If so, can you demonstrate it in a test, since it's rather subtle.

**Review by minestarks — CHANGES_REQUESTED** [review]

**Comment by minestarks** [comment]
@lukas-zech-software rebasing or squashing isn't required but up to you. Please see the review comment about ```ArrayLike<T>``` and revise as necessary. Thank you.

File: jasmine/index.d.ts (hunk 4 above)
Comment (LOZORD) [inline jasmine/index.d.ts]: Sure, SGTM 👍

File: jasmine/index.d.ts [hunk 5]
```diff
@@ -136,9 +143,9 @@ declare namespace jasmine {
     }
 
     interface MatchersUtil {
-        equals(a: any, b: any, customTesters?: Array<CustomEqualityTester>): boolean;
-        contains<T>(haystack: ArrayLike<T> | string, needle: any, customTesters?: Array<CustomEqualityTester>): boolean;
-        buildFailureMessage(matcherName: string, isNot: boolean, actual: any, ...expected: Array<any>): string;
+        equals(a: any, b: any, customTesters?: CustomEqualityTester[]): boolean;
+        contains<T>(haystack: ArrayLike<T> | string, needle: any, customTesters?: CustomEqualityTester[]): boolean;
```
Comment (LOZORD) [inline jasmine/index.d.ts]: Sure; if needed, this can be a follow-up PR.

File: jasminewd2/index.d.ts (hunk 2 above)
Comment (lukas-zech-software) [inline jasminewd2/index.d.ts]: @minestarks @LOZORD 
Seems like I misunderstood the initial question, I thought it was about the syntax used...

I used `ArrayLike` instead of `Array` because it was used in  `jasmine.d.ts` like that.

See https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/jasmine/index.d.ts#L75

I have no particular knowledge of the `jasminewd2/index.d.ts` cited here as I just applied the same changes as I did in the *base* jasmine typings

`ArrayLike` seems to match well with regular arrays like `[1,2,3,]` as you can se in these tests: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/14875/files#diff-443087a7bbdcb27d0da21ebf1b396e9fR108

```
const arr:ArrayLike<number> = [1, 2, 3];  // seems to be valid
```

Comment (lukas-zech-software) [inline jasminewd2/index.d.ts]: If you have concern about using `ArrayLike` instead of `Array` I can investigate why this was initially used and replace it in the base jasmine typings as well

Comment (lukas-zech-software) [inline jasminewd2/index.d.ts]: The code would still look like this with regular arrays though

```
interface ArrayMatchers<T extends Array<T>> extends Matchers<T> {
```

As I want to constrain the type `T` to be compatible with `Array<T>` or is there some error I don`t see?

Comment (LOZORD) [inline jasminewd2/index.d.ts]: I am fine with `ArrayLike` over `Array` (as `Array`/`[]` __should be/implement__ `ArrayLike`).

I was confused about `T extends ArrayLike<T>`. From my understanding that looks like something that would be (could only be) an array of arrays.

For example, let's say you have something like `MyIntList`. That would satisfy `U extends ArrayLike<T>`: `MyIntList extends ArrayLike<number>`.

Comment (minestarks) [inline jasminewd2/index.d.ts]: @lukas-zech-software ```Array``` vs ```ArrayLike``` doesn't matter, that's not the issue. We are suggesting that you drop the ```T extends ArrayLike<T>``` constraint. This is what you would end up with:

```ts
interface ArrayLikeMatchers<T> extends Matchers<ArrayLike<T>> {
 toBe(expected: Expected<ArrayLike<T>>, expectationFailOutput?: any): Promise<void>;
 toEqual(expected: Expected<ArrayLike<T>>, expectationFailOutput?: any): Promise<void>;
 toContain(expected: T, expectationFailOutput?: any): Promise<void>;
 not: ArrayLikeMatchers<T>;
}
```

If you disagree that this would work, could you demonstrate your use case with a test case? Perhaps you're trying to do something subtle here.

Comment (lukas-zech-software) [inline jasminewd2/index.d.ts]: The [Generic Constraint](https://www.typescriptlang.org/docs/handbook/generics.html#generic-constraints) syntax is a bit misleading here.

From my understanding, the `T` has two different meanings here:
`T extends ArrayLike<T>` expresses that a parameter named `T` must implement an interface called`ArrayLike` wich in turn takes a generic parameter *of the type of `T`*

So the first `T` declares the name of the parameter while the second usage in `ArrayLike<T>` uses the type of this parameter like so `ArrayLike<typeof T>`


The usage in the tests seem to prove that, but maybe I got some error in my understanding of how this works?

How else would I express that the generic parameter `T` must be of type `ArrayLike<T>/Array<T>`

Comment (lukas-zech-software) [inline jasminewd2/index.d.ts]: Okay, now I got it. 🛁 

Sorry I had a huge knot in my thinking and could not see past it 😅 

Of course you are correct, I will fix this right away

**Comment by lukas-zech-software** [comment]
I now had to [weaken the typings](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/14875/commits/9b6db1a1561a26ba5644587e1dad420645d112f1#diff-2281b1909cb31985905bfeb1234f16deR244) in `jasmine-jquery` as they have an incompatible extension of `toContain` that expects the same type as `toEqual` which is illogical

**Comment by LOZORD** [comment]
No worries -- jQuery is kind of a quagmire when it comes to types in my experience.

File: jasminewd2/index.d.ts (hunk 2 above)
Comment (minestarks) [inline jasminewd2/index.d.ts]: @lukas-zech-software :) Thank you!

**Review by minestarks — APPROVED** [review]

**Comment by lukas-zech-software** [comment]
@sinedied 

Well it looks like `getParagraphText` [returns a promise](http://www.protractortest.org/#/api?view=webdriver.WebElement.prototype.getText)

So your test tries to compare a Promise with a string, which the new typings show as error.

`jasmine.expect` will not resolve the promise before running its matchers so it currently seems to me as the error is quite correct.

What would you expect the code to do with the promise?

**Comment by sinedied** [comment]
This code works perfectly with protractor, as protractor have extra code to wait for the promise to resolve before trying to run the matcher. Maybe protractor needs to extends the typings on their end, but I bet it can't end well if it's forced on the base jasmine typings.

**Comment by lukas-zech-software** [comment]
If protractor overrides the default behaviour of jasmine, you will need to get typings for this behaviour.

This is nothing the typings of jasmine can take care of, nor should they as this would influence all users of jasmine, no matter if they use protractor or not

And there is nothing on these typings which blocks extending jasmine, as this already work quite fine for other jasmine extension like `jasmine-query` or `jasminewd2` and others

**Comment by lukas-zech-software** [comment]
Btw, you can effectively disable the effect of the new typings per statement by explicitly providing `any` as generic parameter like so
```typescript
 expect<any>(page.getParagraphText()).toEqual('Hello world !');
```

**Comment by sinedied** [comment]
The protractor team will take the point: https://github.com/angular/protractor/issues/4176

**Comment by lorenzodallavecchia** [comment]
I'm experiencing the same issue that @sinedied reported, but with custom matchers that I am developing for our team. The weakening of jQuery types that @lukas-zech-software mentioned may also be an indication of the same issue.

After investigating, it looks like the problem is either caused by Jasmine typings or is a TypeScript limitation.

Our matchers, like the Protractor ones, work on `Promise<T>` (and also on `Observable<T>`).
Both these types can be seen as "wrapper" types, in the same way that arrays are.
The matchers are implemented so that the values of type `T` are unwrapped (i.e. the promise is awaited), and that means the actual matches will take place on `T` instead of the wrapper type.

Example usage:
```typescript
let p = Promise.resolve("foo"); // this is Promise<string>
expect(p).toResolveWith("foo"); // the matcher requires string
```

The typings for our matchers are declared like this:
```typescript
declare function expect<T>(actual: Promise<T>): PromiseMatchers<T>;
declare interface PromiseMatchers<T> {
  toResolveWith(result: T): void;
}
```

All the above does not work because **the `expect` call is being resolved to the wrong overload**.
The Jasmine typings declare the following overloads for `expect`:

```typescript
declare function expect(spy: Function): jasmine.Matchers<any>;
declare function expect<T>(actual: ArrayLike<T>): jasmine.ArrayLikeMatchers<T>;
declare function expect<T>(actual: T): jasmine.Matchers<T>;
```

The third one is capturing any usage of `expect` that is not on a function or on an `ArrayLike`, and the result is that it is not possible to develop a type-safe matcher that works on other "container" types, including Promise, Observable, all iterables and probably more.
In other words, thrid parties may want to do what the second overload is doing with `ArrayLike<T>`.

Now, while I understand the problem, I am not sure about what could be the solution.
It turns out TypeScript is not choosing the "most specific" overload, but just "the first compatible one". This might be a TypeScript limitation.
Is there a way to work around that?

**Comment by lukas-zech-software** [comment]
@lorenzodallavecchia 
The only workaround I found for this is redeclaring the expect function in your TS code.
In any of your **.ts** files (not **d.ts**!) you can redeclare the `expect` functions and add your own in an higher order:

```typescript
// An .ts file included in your buid
declare function expect(spy: Function): jasmine.Matchers<any>;
declare function expect<T>(actual: ArrayLike<T>): jasmine.ArrayLikeMatchers<T>;
declare function expect<T>(actual: Promise<T>): PromiseMatchers<T>;
declare function expect<T>(actual: T): jasmine.Matchers<T>;

interface PromiseMatchers<T> {
  toResolveWith(result: T): void;
}
```

This will effectively overwrite all declaration of the `expect` function present in any `.d.ts` files as the `.ts` files are always processed after the pure declaration files.


Downside to this is, that **all** previously declared definitions of the `expect` functions are overwritten and you'll have to redeclare all you want to use in your code.


This is definitely only a workaround which might help you in short term but I highly recommend to post your use case as a question/bug in the TypeScript repository and see if it gets a proper solution there

**Comment by lorenzodallavecchia** [comment]
Thanks, @lukas-zech-software. I had discovered that solution but ruled it out because of the reasons that you pointed out.
Since we are still using our library in a controlled environment (a couple of projects), I wrote an npm script that modifies the `@types/jasmine` .d.ts file after installation. It is not clean, but it gets the work done.

**Comment by sinedied** [comment]
@lorenzodallavecchia a proper solution was given in the protractor issue: just install `@types/jasminewd2` and you're good.
